blockprod/src/detail/mod.rs

```diff
@@ -47,158 +67,516 @@ impl BlockProduction {
+
         let block_production = Self {
             chain_config,
             chainstate_handle,
             mempool_handle,
             time_getter,
-            builder_tx,
+            job_manager,
+            mining_thread_pool,
         };
+
         Ok(block_production)
     }
-}
 
-#[async_trait::async_trait]
-impl BlockProductionInterface for BlockProduction {
-    fn stop(&self) -> Result<(), BlockProductionError> {
-        self.builder_tx
-            .send(BlockBuilderControlCommand::Stop)
-            .map_err(|_| BlockProductionError::BlockBuilderChannelClosed)?;
-        Ok(())
+    pub fn time_getter(&self) -> &TimeGetter {
+        &self.time_getter
     }
 
-    fn start(&self) -> Result<(), BlockProductionError> {
-        self.builder_tx
-            .send(BlockBuilderControlCommand::Start)
-            .map_err(|_| BlockProductionError::BlockBuilderChannelClosed)?;
-        Ok(())
+    pub async fn stop_all_jobs(&mut self) -> Result<usize, BlockProductionError> {
+        self.job_manager
+            .stop_all_jobs()
+            .await
+            .map_err(BlockProductionError::JobManagerError)
     }
 
-    fn is_connected(&self) -> bool {
-        !self.builder_tx.is_closed()
+    pub async fn stop_job(&mut self, job_key: JobKey) -> Result<bool, BlockProductionError> {
+        Ok(self.job_manager.stop_job(job_key).await? == 1)
     }
 
-    async fn generate_block(
+    pub async fn collect_transactions(
         &self,
-        reward_destination: Destination,
-        transactions: Vec<SignedTransaction>,
-    ) -> Result<Block, BlockProductionError> {
-        let (current_tip_id, current_tip_height) = self
+    ) -> Result<Box<dyn TransactionAccumulator>, BlockProductionError> {
+        let max_block_size = self.chain_config.max_block_size_from_txs();
+        let returned_accumulator = self
+            .mempool_handle
+            .call(move |mempool| {
+                mempool.collect_txs(Box::new(DefaultTxAccumulator::new(max_block_size)))
+            })
+            .await?
+            .map_err(|_| BlockProductionError::MempoolChannelClosed)?;
+        Ok(returned_accumulator)
+    }
+
+    async fn stop_job_in_scope(&mut self, job_key: JobKey, job_key_destroyed: &AtomicBool) {
+        // TODO: this function has to go.
+        // I couldn't find a way to use RAII to call an async function that takes a mut reference to self.
+        // Once this solution is found, this has to go
+        {
+            assert!(
+                !job_key_destroyed.load(std::sync::atomic::Ordering::SeqCst),
+                "Must be true as it was done already"
+            );
+            let result_receiver = self.job_manager.stop_job(job_key.clone());
+            let _stop_result = result_receiver.await;
+        }
+        // We consume the value so that it doesn't happen again
+        job_key_destroyed.store(true, std::sync::atomic::Ordering::SeqCst);
+    }
```

Comment:
I've gone over this several times, and I think because you're calling this in both branches in the below `select!`, then this should be ok.

Reply:
Yes, it works now fine because it's always called before returning. But again, this is naive C way of doing things. It cannot be guaranteed in every case, which can cause memory leaks. Until we have a better solution... we'll see.